### PR TITLE
Rework new issue API

### PR DESCRIPTION
**Pull request by henryju**
* Raw issues are directly written to protobuf report instead of persisit
* Introduce primaryLocation in new issue API

### CODE REVIEW

File: sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java [hunk 1]
```diff
@@ -62,16 +69,18 @@
   private final IssueChangeContext changeContext;
   private final ActiveRules activeRules;
   private final InputPathCache inputPathCache;
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java]: MAJOR Remove this unused "inputPathCache" private field. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)

File: sonar-batch/src/main/java/org/sonar/batch/report/CoveragePublisher.java [hunk 2]
```diff
@@ -54,37 +54,37 @@ public void publish(BatchReportWriter writer) {
...
       });
-      applyLineMeasure(resource.key(), ((InputFile) resource.inputPath()).lines(), CoreMetrics.IT_COVERED_CONDITIONS_BY_LINE_KEY, coveragePerLine, new MeasureOperation() {
+      applyLineMeasure(resource.key(), ((InputFile) resource.inputComponent()).lines(), CoreMetrics.IT_COVERED_CONDITIONS_BY_LINE_KEY, coveragePerLine, new MeasureOperation() {
         @Override
         public void apply(String value, Coverage.Builder builder) {
           builder.setItCoveredConditions(Integer.parseInt(value));
         }
       });
-      applyLineMeasure(resource.key(), ((InputFile) resource.inputPath()).lines(), CoreMetrics.OVERALL_COVERED_CONDITIONS_BY_LINE_KEY, coveragePerLine, new MeasureOperation() {
+      applyLineMeasure(resource.key(), ((InputFile) resource.inputComponent()).lines(), CoreMetrics.OVERALL_COVERED_CONDITIONS_BY_LINE_KEY, coveragePerLine, new MeasureOperation() {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/report/CoveragePublisher.java]: MINOR Split this 181 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultInputModule.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.api.batch.fs.internal;
+
+import org.sonar.api.batch.fs.InputModule;
+
+/**
+ * @since 5.2
+ */
+public class DefaultInputModule extends DefaultInputComponent implements InputModule {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultInputModule.java]: MAJOR Override this superclass' "equals" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2160)

File: sonar-batch/src/main/java/org/sonar/batch/issue/ModuleIssues.java [hunk 4]
```diff
@@ -37,55 +46,126 @@
...
     }
     return false;
   }
 
-  private static void validateRule(DefaultIssue issue, @Nullable Rule rule) {
+  private void applyAdditionalLocations(Issue issue) {
+    for (org.sonar.api.batch.sensor.issue.IssueLocation additionalLocation : issue.locations()) {
+      locationBuilder.clear();
+      String locationComponentKey = ((DefaultInputComponent) additionalLocation.inputComponent()).key();
+      locationBuilder.setComponentRef(componentCache.get(locationComponentKey).batchId());
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/ModuleIssues.java]: BLOCKER NullPointerException might be thrown as 'get' is nullable here [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)

File: sonar-batch/src/main/java/org/sonar/batch/issue/ModuleIssues.java [hunk 5]
```diff
@@ -37,55 +46,126 @@
...
+    }
+  }
+
+  private void applyExecutionFlows(Issue issue) {
+    for (ExecutionFlow executionFlow : issue.executionFlows()) {
+      flowBuilder.clear();
+      for (org.sonar.api.batch.sensor.issue.IssueLocation location : executionFlow.locations()) {
+        locationBuilder.clear();
+        String locationComponentKey = ((DefaultInputComponent) location.inputComponent()).key();
+        locationBuilder.setComponentRef(componentCache.get(locationComponentKey).batchId());
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/ModuleIssues.java]: BLOCKER NullPointerException might be thrown as 'get' is nullable here [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)

File: sonar-batch/src/main/java/org/sonar/batch/issue/ModuleIssues.java [hunk 6]
```diff
@@ -37,55 +46,126 @@
...
+    locationBuilder.clear();
+    // non-null fields
+    builder.setSeverity(severity);
+    builder.setRuleRepository(issue.ruleKey().repository());
+    builder.setRuleKey(issue.ruleKey().rule());
+    builder.setAttributes(KeyValueFormat.format(issue.attributes()));
+    builder.setMsg(primaryMessage);
+    locationBuilder.setMsg(primaryMessage);
+
+    locationBuilder.setComponentRef(component.batchId());
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/ModuleIssues.java]: BLOCKER NullPointerException might be thrown as 'component' is nullable here [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)

File: sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java [hunk 7]
```diff
@@ -133,16 +139,14 @@
   boolean addIssue(Issue issue);
 
   /**
-   * Unresolved issues, including the issues reported by end-users.
-   * <p/>
-   * {@link org.sonar.api.batch.Decorator}s calling this method must be annotated with {@code @DependsUpon(DecoratorBarriers.ISSUES_TRACKED)}.
+   * @deprecated since 5.2 no more decorators on batch side
    */
+  @Deprecated
   List<Issue> issues();
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java [hunk 8]
```diff
@@ -133,16 +139,14 @@
...
 
   /**
-   * Issues marked as resolved during this scan.
-   * <p/>
-   * {@link org.sonar.api.batch.Decorator}s calling this method must be annotated with {@code @DependsUpon(DecoratorBarriers.ISSUES_TRACKED)}.
+   * @deprecated since 5.2 no more decorators on batch side
    */
+  @Deprecated
   List<Issue> resolvedIssues();
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: sonar-batch/src/main/java/org/sonar/batch/issue/IssueFilters.java [hunk 9]
```diff
@@ -59,4 +73,145 @@ public boolean accept(DefaultIssue issue) {
       return false;
     }
   }
+
+  private Issue toIssueForIssueFilter(final String componentKey, final BatchReport.Issue rawIssue) {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/IssueFilters.java]: MAJOR The Cyclomatic Complexity of this method "toIssueForIssueFilter" is 42 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR This method has 140 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS138)

File: sonar-batch/src/main/java/org/sonar/batch/issue/IssueFilters.java [hunk 10]
```diff
@@ -59,4 +73,145 @@ public boolean accept(DefaultIssue issue) {
...
+    return new Issue() {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/IssueFilters.java]: MAJOR Reduce this anonymous class number of lines from 137 to at most 20, or make it a named class. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1188)

File: sonar-batch/src/main/java/org/sonar/batch/mediumtest/TaskResult.java [hunk 11]
```diff
@@ -121,10 +123,27 @@ private void storeFs(ProjectScanContainer container) {
...
+      while (it.hasNext()) {
+        result.add(it.next());
+      }
+    } catch (Exception e) {
+      throw new IllegalStateException("Can't read issues for " + inputPath.absolutePath(), e);
+    }
+    return result;
+  }
+
+  private String key(InputPath inputPath) {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/mediumtest/TaskResult.java]: MAJOR Make "key" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 19 issues:
* BLOCKER 3 blocker
* CRITICAL 4 critical
* MAJOR 9 major
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "createIssues" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonarqube/blob/3a7b88081c938323a77044d10122acbdecef50b4/plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/rule/MultilineIssuesSensor.java#L64) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/3a7b88081c938323a77044d10122acbdecef50b4/sonar-batch/src/main/java/org/sonar/batch/issue/tracking/ServerIssueRepository.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/3a7b88081c938323a77044d10122acbdecef50b4/sonar-batch/src/main/java/org/sonar/batch/scan/report/SourceProvider.java#L40) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* CRITICAL [Method 'public java.nio.file.Path path()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/3a7b88081c938323a77044d10122acbdecef50b4/sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputPath.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public java.lang.Integer line()' has been removed](https://github.com/SonarSource/sonarqube/blob/3a7b88081c938323a77044d10122acbdecef50b4/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.batch.fs.InputPath inputPath()' has been removed](https://github.com/SonarSource/sonarqube/blob/3a7b88081c938323a77044d10122acbdecef50b4/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.batch.sensor.issue.NewIssueLocation newLocation()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/3a7b88081c938323a77044d10122acbdecef50b4/sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)

**Comment by SonarTech** [comment]
SonarQube analysis reported 16 issues:
* BLOCKER 3 blocker
* CRITICAL 4 critical
* MAJOR 7 major
* INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "createIssues" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonarqube/blob/772fe9c8f9b3e07b88a5d06534bc308fdc715265/plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/rule/MultilineIssuesSensor.java#L64) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/772fe9c8f9b3e07b88a5d06534bc308fdc715265/sonar-batch/src/main/java/org/sonar/batch/issue/tracking/ServerIssueRepository.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/772fe9c8f9b3e07b88a5d06534bc308fdc715265/sonar-batch/src/main/java/org/sonar/batch/scan/report/SourceProvider.java#L40) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* CRITICAL [Method 'public java.nio.file.Path path()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/772fe9c8f9b3e07b88a5d06534bc308fdc715265/sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputPath.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public java.util.List locations()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/772fe9c8f9b3e07b88a5d06534bc308fdc715265/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.batch.rule.Severity overriddenSeverity()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/772fe9c8f9b3e07b88a5d06534bc308fdc715265/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.issue.Issuable$IssueBuilder addExecutionFlow(org.sonar.api.batch.sensor.issue.NewIssueLocation[])' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/772fe9c8f9b3e07b88a5d06534bc308fdc715265/sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)

**Comment by simonbrandhof** [comment]
Many changes in this PR, so I did only a quick review. LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 16 issues:
* BLOCKER 3 blocker
* CRITICAL 4 critical
* MAJOR 7 major
* INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "createIssues" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonarqube/blob/1cdfef4c0652a40e16fbe76b58eec76c35c6b634/plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/rule/MultilineIssuesSensor.java#L64) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/1cdfef4c0652a40e16fbe76b58eec76c35c6b634/sonar-batch/src/main/java/org/sonar/batch/issue/tracking/ServerIssueRepository.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/1cdfef4c0652a40e16fbe76b58eec76c35c6b634/sonar-batch/src/main/java/org/sonar/batch/scan/report/SourceProvider.java#L40) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* CRITICAL [Method 'public java.nio.file.Path path()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/1cdfef4c0652a40e16fbe76b58eec76c35c6b634/sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputPath.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public java.util.Map attributes()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/1cdfef4c0652a40e16fbe76b58eec76c35c6b634/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public java.lang.String message()' has been removed](https://github.com/SonarSource/sonarqube/blob/1cdfef4c0652a40e16fbe76b58eec76c35c6b634/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.batch.sensor.issue.NewIssueLocation newLocation()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/1cdfef4c0652a40e16fbe76b58eec76c35c6b634/sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)

**Comment by SonarTech** [comment]
SonarQube analysis reported 16 issues:
* BLOCKER 3 blocker
* CRITICAL 4 critical
* MAJOR 7 major
* INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "createIssues" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonarqube/blob/169422a87ae36d67760793f9ebff1c1e450439f9/plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/rule/MultilineIssuesSensor.java#L64) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/169422a87ae36d67760793f9ebff1c1e450439f9/sonar-batch/src/main/java/org/sonar/batch/issue/tracking/ServerIssueRepository.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/169422a87ae36d67760793f9ebff1c1e450439f9/sonar-batch/src/main/java/org/sonar/batch/scan/report/SourceProvider.java#L40) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* CRITICAL [Method 'public java.nio.file.Path path()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/169422a87ae36d67760793f9ebff1c1e450439f9/sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputPath.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public java.lang.String message()' has been removed](https://github.com/SonarSource/sonarqube/blob/169422a87ae36d67760793f9ebff1c1e450439f9/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.batch.rule.Severity overriddenSeverity()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/169422a87ae36d67760793f9ebff1c1e450439f9/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.issue.Issuable$IssueBuilder addLocation(org.sonar.api.batch.sensor.issue.NewIssueLocation)' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/169422a87ae36d67760793f9ebff1c1e450439f9/sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)

**Comment by SonarTech** [comment]
SonarQube analysis reported 16 issues:
* BLOCKER 3 blocker
* CRITICAL 4 critical
* MAJOR 7 major
* INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "createIssues" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonarqube/blob/318bc5d9ce9995a3e4bfd89dac51fb60aa03c567/plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/rule/MultilineIssuesSensor.java#L64) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/318bc5d9ce9995a3e4bfd89dac51fb60aa03c567/sonar-batch/src/main/java/org/sonar/batch/issue/tracking/ServerIssueRepository.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* MAJOR [Remove this unused "inputPathCache" private field.](https://github.com/SonarSource/sonarqube/blob/318bc5d9ce9995a3e4bfd89dac51fb60aa03c567/sonar-batch/src/main/java/org/sonar/batch/scan/report/SourceProvider.java#L40) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* CRITICAL [Method 'public java.nio.file.Path path()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/318bc5d9ce9995a3e4bfd89dac51fb60aa03c567/sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputPath.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.batch.sensor.issue.IssueLocation primaryLocation()' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/318bc5d9ce9995a3e4bfd89dac51fb60aa03c567/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public java.lang.String severity()' has been removed](https://github.com/SonarSource/sonarqube/blob/318bc5d9ce9995a3e4bfd89dac51fb60aa03c567/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/issue/Issue.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)
* CRITICAL [Method 'public org.sonar.api.issue.Issuable$IssueBuilder addExecutionFlow(org.sonar.api.batch.sensor.issue.NewIssueLocation[])' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/318bc5d9ce9995a3e4bfd89dac51fb60aa03c567/sonar-plugin-api/src/main/java/org/sonar/api/issue/Issuable.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)

**Pull request merged**